### DISCUSSION TITLE
relaxationfactor

### DISCUSSION THREAD

**Discussion by bigfooted** [su2code/SU2#1244] [Q&A]
We now have the function ComputeUnderRelaxationFactor. I think this has a lot of potential to improve convergence for many cases but there is currently no user interaction. I propose to make this more accessible to users by introducing config options for all transported variables, but I have some questions, I hope somebody knows the answer.

1. why is it active for SA and not for SST? Is this based on experience with some testcases? I think in general we should be able to activate it for any transported variable.
2. why is it active for compressible but not incompressible? In CIncEuler::CompleteIteration we have  CompleteImplicitIteration_impl<false> so not active for incompressible. For some cases, activating this for incompressible makes the difference between convergence and nonconvergence.
3. Related to point 2 is that in CFVMFlowSolverBase<V, R>::ComputeUnderRelaxationFactor we only use it for iVar==0 and iVar==nVar-1 (density and energy for compressible). Why is that? I think in general we should underrelax all variables, although there are probably many cases where underrelaxing energy for compressible cases has the biggest impact.

**Comment by WallyMaier** [comment]
@bigfooted I just stumbled across this as well while writing the implicit functions for NEMO.  I had the same questions as you, particularly due to the NEMO solvers having different variable indices.  Id be happy to help out here.

**Reply by bigfooted** [reply]
Hi Wally, What I understood from @pcarruscag is that the variable under consideration is not allowed to become zero. For SA, nu_{tilde} cannot be zero, but for SST, k can become zero.  In CFVMFlowSolverBase.inl we have for instance:
`su2double ratio = fabs(LinSysSol[index]) / (fabs(nodes->GetSolution(iPoint, iVar)) + EPS);`

So although division by zero is captured, I guess that this is not sufficient to have good performance?

**Reply by pcarruscag** [reply]
Just to make sure that credit goes where it is due. That is the explanation @economon gave me when I had similar questions.
Tom implemented the automatic underrelaxation algorithm and associated CFL adaptation strategy in #790.
It is all explained in https://arc.aiaa.org/doi/abs/10.2514/6.2021-0855, the references therein might be useful.
